host.go

```diff
@@ -108,40 +128,117 @@ func (h *gpbftRunner) ValidateMessage(msg *gpbft.GMessage) (gpbft.ValidatedMessa
 // The chain should be a suffix of the last chain notified to the host via
 // ReceiveDecision (or known to be final via some other channel).
 func (h *gpbftHost) GetProposalForInstance(instance uint64) (*gpbft.SupplementalData, gpbft.ECChain, error) {
-	// TODO: this is just a complete fake
+	var baseTsk gpbft.TipSetKey
+	if instance == 0 {
+		ts, err := h.client.ec.GetTipsetByEpoch(h.runningCtx,
+			h.manifest.BootstrapEpoch-finality)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("getting boostrap base: %w", err)
+		}
+		baseTsk = ts.Key()
+	} else {
+		cert, err := h.client.certstore.Get(h.runningCtx, instance-1)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("getting cert for previous instance(%d): %w", instance-1, err)
+		}
+		baseTsk = cert.ECChain.Head().Key
+	}
 
-	pt, _, err := h.GetCommitteeForInstance(0)
+	baseTs, err := h.client.ec.GetTipset(h.runningCtx, baseTsk)
 	if err != nil {
-		return nil, nil, xerrors.Errorf("getting power table: %w", err)
+		return nil, nil, xerrors.Errorf("getting base TS: %w", err)
+	}
+	headTs, err := h.client.ec.GetHead(h.runningCtx)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("getting head TS: %w", err)
+	}
+
+	collectedChain, err := h.collectChain(baseTs, headTs)
```

Comment:
This could be an issue if we fall really behind in gpbft and don't see the finality certificates. We could collect a very large chain.

I wonder if, instead, we should have a way to ask the client for the next N epochs after tipset A, on the chain ending in tipset B.

Reply:
Ah, I see the TODO above now.